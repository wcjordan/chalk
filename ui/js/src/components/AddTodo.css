.add-todo {
  background-color: #98c1d9;
  border: none;
  display: flex;
  font-family: monospace;
  padding: 8px;
  width: 100%;
}

.add-todo-input {
  background-color: inherit;
  border: inherit;
  color: #293241;
  font-family: inherit;
  font-size: 20px;
  line-height: 1.2;
  margin-left: 4px;
  padding: 4px;
  flex-grow: 1;
}

.add-todo-input:focus {
  outline: none;
}

.add-todo-input::placeholder {
  color: #e0fbfc;
}

.add-todo-icon {
  transform: rotate(90deg);
}

.add-todo-icon:before {
  color: #e0fbfc;
  content: '‚ùØ';
  font-size: 32px;
  padding: 0px 4px;
}
