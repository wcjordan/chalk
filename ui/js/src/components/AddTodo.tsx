import React, { useCallback } from 'react';
import {
  StyleSheet,
  TextInput,
  View,
  NativeSyntheticEvent,
  TextInputSubmitEditingEventData,
} from 'react-native';

const styles = StyleSheet.create({
  addTodo: {
    backgroundColor: '#98c1d9',
    borderWidth: 0,
    display: 'flex',
    fontFamily: 'monospace',
    padding: 8,
    width: '100%',
  },
  addTodoInput: {
    backgroundColor: 'transparent',
    color: '#293241',
    fontFamily: 'inherit',
    fontSize: 20,
    marginLeft: 4,
    padding: 4,
    flexGrow: 1,
  },
});

const AddTodo: React.FC<Props> = function (props: Props) {
  const { createTodo } = props;
  const addTodo = useCallback(
    (event: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => {
      createTodo(event.nativeEvent.text);
    },
    [createTodo],
  );

  // TODO remove input outline on focus
  return (
    <View style={styles.addTodo}>
      <TextInput
        nativeID="add-todo-input"
        style={styles.addTodoInput}
        placeholder="Add a new todo..."
        placeholderTextColor="#e0fbfc"
        onSubmitEditing={addTodo}
      />
    </View>
  );
};

type Props = {
  createTodo: (description: string) => void;
};

export default AddTodo;
