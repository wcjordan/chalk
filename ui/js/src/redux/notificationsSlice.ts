import { createSlice } from '@reduxjs/toolkit';

import { NotificationsState } from './types';

const initialState: NotificationsState = {
  notificationQueue: [],
};
export default createSlice({
  name: 'notifications',
  initialState,
  reducers: {
    addNotification: (state, action) => {
      state.notificationQueue.push(action.payload);
    },
    dismissNotification: (state) => {
      state.notificationQueue.shift();
    },
  },
});
