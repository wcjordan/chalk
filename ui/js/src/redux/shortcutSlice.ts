import { createSlice } from '@reduxjs/toolkit';

import { ShortcutState } from './types';

const initialState: ShortcutState = {
  operations: [],
  latestGeneration: 0,
};
export default createSlice({
  name: 'shortcuts',
  initialState,
  reducers: {
    addEditTodoOperation: (state, action) => {
      state.operations.push({
        type: 'EDIT_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    addMoveTodoOperation: (state, action) => {
      state.operations.push({
        type: 'MOVE_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    incrementGenerations: (state) => {
      state.latestGeneration = state.latestGeneration + 1;
    },
    clearOperationsUpThroughGeneration: (state, action) => {
      state.operations = state.operations.filter(
        (op) => op.generation > action.payload,
      );
    },
  },
});
