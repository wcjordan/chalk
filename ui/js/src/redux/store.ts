import { configureStore } from '@reduxjs/toolkit';
import rootReducer, { listTodos } from './reducers';

const store = configureStore({
  reducer: rootReducer,
});

// TODO dispose
window.setInterval(() => store.dispatch(listTodos()), 1000);

export default store;
