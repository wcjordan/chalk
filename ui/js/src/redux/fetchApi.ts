import Cookies from 'js-cookie';
import { Platform } from 'react-native';

import { getEnvFlags } from '../helpers';
import { RootState } from './store';

const webCsrfToken = Cookies.get('csrftoken') as string;
export function getCsrfToken(getState: () => RootState): string {
  if (Platform.OS === 'web') {
    return webCsrfToken;
  }
  return getState().workspace.csrfToken as string;
}

export async function list<T>(apiUri: string): Promise<Array<T>> {
  const response = await fetch(apiUri, getRequestOpts('GET'));
  return handleResponse<Array<T>>(response);
}

export async function patch<T, Q extends { id: number }>(
  apiUri: string,
  entryPatch: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('PATCH', csrfToken);
  requestOpts.body = JSON.stringify(entryPatch);
  const response = await fetch(`${apiUri}${entryPatch.id}/`, requestOpts);
  return handleResponse<T>(response);
}

export async function create<T, Q>(
  apiUri: string,
  newEntry: Q,
  csrfToken: string,
): Promise<T> {
  return postRequest<T, Q>(apiUri, newEntry, csrfToken);
}

export async function postRequest<T, Q>(
  uri: string,
  data: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = JSON.stringify(data);
  const response = await fetch(uri, requestOpts);
  return handleResponse<T>(response);
}

function getRequestOpts(
  method: string,
  csrfToken: string | null = null,
): RequestInit {
  const headers: HeadersInit = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
  if (csrfToken) {
    headers['X-CSRFToken'] = csrfToken;
  }

  const credentials: RequestCredentials = Platform.select({
    native: 'include',
    default: 'same-origin',
  });

  return {
    credentials: credentials,
    headers: headers,
    method: method,
  };
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(`${response.status} ${response.statusText}`);
  }
  return await response.json();
}

export function getWsRoot(): string {
  if (Platform.OS === 'web') {
    return '';
  }

  const subdomain =
    getEnvFlags().ENVIRONMENT === 'prod' ? 'chalk' : 'chalk-dev';
  return `http://${subdomain}.flipperkid.com/`
}

// Used to exchange login token for session cookie in mobile login flow
export async function completeAuthCallback(token: string) {
  return await fetch(
    `${getWsRoot()}api/todos/auth_callback/?code=${token}`,
    getRequestOpts('GET'),
  );
}

export async function logSessionData(
  dataString: string,
  csrfToken: string,
): Promise<string> {
  
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = dataString;
  const response = await fetch(
    `${getWsRoot()}api/todos/log_session_data/`,
    requestOpts,
  );
  return handleResponse<string>(response);
}
