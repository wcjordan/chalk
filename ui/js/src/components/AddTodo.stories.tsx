import React from 'react';
import { Provider } from 'react-redux';
import { setupStore } from '../redux/store';
import AddTodo from './AddTodo';

const wrapper = (component) => (
  <Provider store={setupStore()}>{component}</Provider>
);

export default {
  title: 'Add Todo',
  component: AddTodo,
};
export const DefaultAddTodo: React.FC = () => wrapper(<AddTodo />);
