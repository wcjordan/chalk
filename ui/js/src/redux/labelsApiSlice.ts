import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ApiState, ReduxState, Label } from './types';
import { getWsRoot, list } from './fetchApi';

const API_NAME = 'labelsApi';

// Exported for testing
export function getLabelsApi(): string {
  return `${getWsRoot()}api/todos/labels/`;
}

const initialState: ApiState<Label> = {
  entries: [],
  initialLoad: true,
  loading: false,
};

export const listLabels = createAsyncThunk<
  Label[],
  void,
  { state: ReduxState }
>(`${API_NAME}/list`, async () => list<Label>(getLabelsApi()), {
  condition: (_unused, { getState }) => {
    return !getState().labelsApi.loading;
  },
});

export default createSlice({
  name: API_NAME,
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(listLabels.pending, (state) => {
        state.loading = true;
      })
      .addCase(listLabels.fulfilled, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        state.entries = action.payload;
      })
      .addCase(listLabels.rejected, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        console.warn(`Loading Labels failed. ${action.error.message}`);
      });
  },
});
