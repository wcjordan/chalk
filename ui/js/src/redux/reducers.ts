import { ThunkAction } from 'redux-thunk';
import { Action } from '@reduxjs/toolkit';

import { ReduxState, TodoPatch } from './types';
import labelsApiSlice, { listLabels } from './labelsApiSlice';
import notificationsSlice from './notificationsSlice';
import shortcutSlice from './shortcutSlice';
import todosApiSlice, {
  createTodo,
  listTodos,
  updateTodo as updateTodoApi,
} from './todosApiSlice';
import workspaceSlice from './workspaceSlice';
import { completeAuthCallback } from './fetchApi';

type AppThunk = ThunkAction<void, ReduxState, unknown, Action<string>>;

export const updateTodo =
  (todoPatch: TodoPatch): AppThunk =>
  (dispatch) => {
    return Promise.all([
      dispatch(
        notificationsSlice.actions.addNotification(
          `Saving Todo: ${todoPatch.description}`,
        ),
      ),
      dispatch(workspaceSlice.actions.setEditTodoId(null)),
      dispatch(shortcutSlice.actions.addEditTodoOperation(todoPatch)),
      dispatch(updateTodoApi(todoPatch)),
    ]);
  };

export const updateTodoLabels =
  (labels: string[]): AppThunk =>
  (dispatch, getState) => {
    const todoId = getState().workspace.labelTodoId;
    if (todoId === null) {
      throw new Error('Unable to edit a todo w/ null ID');
    }

    return dispatch(
      updateTodoApi({
        id: todoId,
        labels,
      }),
    );
  };

// Used to exchange login token for session cookie in mobile login flow
export const completeAuthentication =
  (accessToken: string): AppThunk =>
  async (dispatch) => {
    try {
      const response = await completeAuthCallback(accessToken);
      const cookies = response.headers.get('set-cookie')?.split(' ');
      const csrfToken = cookies
        ?.find((val) => val.startsWith('csrftoken'))
        ?.slice(10, -1);

      if (response.status !== 200) {
        const responseText = await response.text();
        const message = `Login failed with status: ${response.status} ${response.statusText}\n${responseText}`;
        dispatch(notificationsSlice.actions.addNotification(message));
        throw new Error(message);
      }

      if (!csrfToken) {
        dispatch(
          notificationsSlice.actions.addNotification(
            'Unexpectedly missing CSRFToken.  Please refresh and login again.',
          ),
        );
        throw new Error(
          `CSRFToken missing from set-cookie header\n${response.headers.get(
            'set-cookie',
          )}`,
        );
      }

      dispatch(
        workspaceSlice.actions.logIn({
          loggedIn: true,
          csrfToken: csrfToken,
        }),
      );
    } catch (ex) {
      dispatch(
        notificationsSlice.actions.addNotification(
          'Login failed with a network error.  Please refresh and login again.',
        ),
      );
      throw ex;
    }
  };

export const addNotification = notificationsSlice.actions.addNotification;
export const dismissNotification =
  notificationsSlice.actions.dismissNotification;
export const setEditTodoId = workspaceSlice.actions.setEditTodoId;
export const setLabelTodoId = workspaceSlice.actions.setLabelTodoId;
export const setWorkContext = workspaceSlice.actions.setWorkContext;
export const toggleLabel = workspaceSlice.actions.toggleLabel;
export const toggleShowCompletedTodos =
  workspaceSlice.actions.toggleShowCompletedTodos;
export const toggleShowLabelFilter =
  workspaceSlice.actions.toggleShowLabelFilter;
export { createTodo, listLabels, listTodos };
export default {
  labelsApi: labelsApiSlice.reducer,
  notifications: notificationsSlice.reducer,
  shortcuts: shortcutSlice.reducer,
  todosApi: todosApiSlice.reducer,
  workspace: workspaceSlice.reducer,
};
