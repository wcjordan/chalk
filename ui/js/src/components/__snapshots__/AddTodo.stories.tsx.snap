// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Todo DefaultAddTodo smoke-test 1`] = `
<div class="css-g5y9jx r-13awgt0">
  <div class="css-g5y9jx r-13awgt0 r-12vffkv">
    <div class="css-g5y9jx r-1kly57o r-13qz1uu">
      <div class="css-g5y9jx"
           style="background-color: rgb(231, 224, 236); border-top-left-radius: 4px; border-top-right-radius: 4px;"
      >
        <div class="css-g5y9jx r-1p0dtai r-1d2f490 r-u8s1d r-zchlnj r-184en5c r-109y4c4"
             data-testid="text-input-underline"
             style="background-color: rgb(73, 69, 79); transform: scaleY(0.5);"
        >
        </div>
        <div class="css-g5y9jx r-16y2uox r-1mdbw0j r-wk8lta"
             style="min-height: 56px;"
        >
          <textarea placeholder="Add a new todo..."
                    autocapitalize="sentences"
                    autocomplete="on"
                    autocorrect="on"
                    dir="auto"
                    rows="3"
                    spellcheck="true"
                    virtualkeyboardpolicy="auto"
                    class="css-11aywtz r-6taxm2 r-16y2uox r-crgep1 r-1x2decy r-xd6kpl r-1m04atk r-1pyaxff r-tskmnb"
                    data-testid="add-todo-input"
                    style="--placeholderTextColor: rgba(73, 69, 79, 1); letter-spacing: 0.15px; line-height: 19.2px; font-size: 16px; color: rgb(28, 27, 31); vertical-align: top; text-align: left; min-width: 65px; outline: none;"
          >
          </textarea>
        </div>
      </div>
    </div>
  </div>
</div>
`;
